package NYT.NChunkHolder.NProto;

import "yt/ytlib/chunk_holder/chunk.proto";
import "yt/ytlib/misc/guid.proto";


////////////////////////////////////////////////////////////////////////////////

message TReqStartChunk
{
    required NYT.NProto.TGuid chunk_id = 1;
}

message TRspStartChunk
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqFinishChunk
{
    required NYT.NProto.TGuid chunk_id = 1;
    required TChunkMeta chunk_meta = 2;
}

message TRspFinishChunk
{
    required TChunkInfo chunk_info = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqPutBlocks
{
    required NYT.NProto.TGuid chunk_id = 1;
    required int32 start_block_index = 2;
}

message TRspPutBlocks
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqSendBlocks
{
    required NYT.NProto.TGuid chunk_id = 1;
    required string address = 2;
    required int32 start_block_index = 3;
    required int32 block_count = 4;
}

message TRspSendBlocks
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqFlushBlock
{
    required NYT.NProto.TGuid chunk_id = 1;
    required int32 block_index = 2;
}

message TRspFlushBlock
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqGetBlocks
{
    required NYT.NProto.TGuid chunk_id = 1;
    repeated int32 block_indexes = 2;
    optional string peer_address = 3;
    optional uint64 peer_expiration_time = 4;
}

message TRspGetBlocks
{
    message TBlockInfo
    {
        required bool data_attached = 1;
        repeated string peer_addresses = 2;
    }

    repeated TBlockInfo blocks = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqPingSession
{
    required NYT.NProto.TGuid chunk_id = 1;
}

message TRspPingSession
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqGetChunkMeta
{
    required NYT.NProto.TGuid chunk_id = 1;
    repeated int32 extension_tags = 2;

    // If true, then extension_tags are ignored and 
    // the whole available meta is returned.
    optional bool full_meta = 3 [default = false];
}

message TRspGetChunkMeta
{
    required TChunkMeta chunk_meta = 1;
}

////////////////////////////////////////////////////////////////////////////////

message TReqPrecacheChunk
{
    required NYT.NProto.TGuid chunk_id = 1;
}

message TRspPrecacheChunk
{ }

////////////////////////////////////////////////////////////////////////////////

message TReqUpdatePeer
{
    message TBlockId
    {
        required NYT.NProto.TGuid chunk_id = 1;
        required int32 block_index = 2;
    }

    repeated TBlockId block_ids = 1;
    required string peer_address = 2;
    required uint64 peer_expiration_time = 3;
}

////////////////////////////////////////////////////////////////////////////////
